### PR TITLE
Fix crash in ajava-based WPI related to captures

### PR DESCRIPTION
**Pull request by kelloggm**
I'm fairly confident that there is a better way to do this than catching the `ClassCastException`, but I was worried of the performance implications of calling `asSuper` on every inference of a field's type, since only very rarely will the two `AnnotatedTypeMirror` not actually have the same actual Java class (causing the exception).

Technically, their `kind` is the same - they are both typevars - so it cannot be checked that way, even though the underlying Java class of one is `AnnotatedDeclaredType`. I think it's probably bad that `updateAtmWithLub` directly casts to `AnnotatedTypeVariable`, but I'm not sure how else it could work.

### CODE REVIEW

**Comment by smillst** [comment]
> Technically, their kind is the same - they are both typevars - so it cannot be checked that way, even though the underlying Java class of one is AnnotatedDeclaredType. I think it's probably bad that updateAtmWithLub directly casts to AnnotatedTypeVariable, but I'm not sure how else it could work.

Are you saying there's an `AnnotatedDelcaredType` that returns `TYPEVAR` when `getKind` is called?  If so, that needs to be fix.

File: framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceImplementation.java [hunk 1]
```diff
@@ -804,7 +804,17 @@ protected void updateAnnotationSet(
 
     AnnotatedTypeMirror atmFromStorage =
         storage.atmFromStorageLocation(rhsATM.getUnderlyingType(), annotationsToUpdate);
-    updateAtmWithLub(rhsATM, atmFromStorage);
+    // It is possible that rhsATM is a type variable but atmFromStorage is not, which
+    // would cause a ClassCastException in updateAtmWithLub(). This can
+    // happen for example when a local variable of type ? extends T is assigned to a
```
Comment (smillst) [inline framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceImplementation.java:809]: A local variable can't have a wildcard type.  I think you when an expression has type `? extends T`.

**Review by smillst — CHANGES_REQUESTED** [review]

**Comment by kelloggm** [comment]
> Are you saying there's an `AnnotatedDelcaredType` that returns `TYPEVAR` when getKind is called? If so, that needs to be fix.

I think that is what's happening. I've added debug code to the `catch` block in this PR so that you can see for yourself (by running `./gradlew ainferIndexAjavaTest -Pemit.test.debug`). That prints out something like this every time the exception is caught:
```
    Catching a ClassCastException: java.lang.ClassCastException: class org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedDeclaredType cannot be cast to class org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedTypeVariable (org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedDeclaredType and org.checkerframework.framework.type.AnnotatedTypeMirror$AnnotatedTypeVariable are in unnamed module of loader 'app')
    The cast must be one of the following two AnnotatedTypeMirrors. The type to cast to is decided by the kind of rhsATM.
    rhsATM: capture#40 extends T extends Object
    rhsATM.getKind(): TYPEVAR
    atmFromStorage: T extends Object
    atmFromStorage.getKind(): TYPEVAR
```

I have no idea where the bad `AnnotatedTypeMirror` is coming from - I'm not sure even where to start looking to debug that. My goal with this PR was to treat the symptom of this problem so that I can run WPI without a crash 🙃  However, I agree this seems like it breaks one of our invariants for the CF as a whole.

File: framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceImplementation.java [hunk 2]
```diff
@@ -804,7 +804,26 @@ protected void updateAnnotationSet(
...
-    updateAtmWithLub(rhsATM, atmFromStorage);
+    // It is possible that rhsATM is a type variable but atmFromStorage is not, which
+    // would cause a ClassCastException in updateAtmWithLub(). This can
+    // happen for example when an expression of type ? extends T is assigned to a
+    // field of type T, such as in the test ainfer-index/non-annotated/Dataset6Crash.java.
+    try {
+      updateAtmWithLub(rhsATM, atmFromStorage);
+    } catch (ClassCastException c) {
+      System.out.println("Catching a ClassCastException: " + c);
+      System.out.println(
```
[the remark this answers is not in the document]
Comment (kelloggm) [reply framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceImplementation.java]: You were exactly right - thanks for figuring this out! I'm glad there wasn't a deeper problem, and I've now written a proper fix.

**Review by smillst — APPROVED** [review]

**Comment by kelloggm** [comment]
@smillst I'll need you to review this again: the tests for the version you approved earlier today failed for a good reason: the check did the wrong thing if one of the arguments to `updateAtmWithLub` was a null type. With this improved handling of type variable bounds in `updateAtmWithLub`, I was also able to remove another workaround for a similar problem that I added a few months ago.

**Review by smillst — APPROVED** [review]

**Pull request merged**